src/EventTeachingService.cpp

```diff
@@ -11,259 +11,427 @@
       {
-        // DEBUG_SERIAL << F("> deleting event at index = ") << index << F(", evs ") << endl;
-        module_config->cleareventEEPROM(index);
-
-        // update hash table
-        module_config->updateEvHashEntry(index);
-
-        // respond with WRACK
-        controller->sendWRACK();
-      } else {
-        // DEBUG_SERIAL << F("> did not find event to unlearn") << endl;
-        // respond with CMDERR
-        controller->sendCMDERR(10);
+        inhibitLearn();
       }
-    } // if in learn mode
 
-    return PROCESSED;
+      return PROCESSED;
+
+    case OPC_RQEVN:
+      // 58 - request for number of stored events
+      // DEBUG_SERIAL << F("> RQEVN -- number of stored events for nn = ") << nn << endl;
 
-  case OPC_NNULN:
-    // received NNULN -- exit from learn mode
+      if (nn == module_config->nodeNum) 
+      {
+        // respond with 0x74 NUMEV
+        controller->sendMessageWithNN(OPC_NUMEV, module_config->numEvents());
+      }
 
-    if (nn == module_config->nodeNum)
-    {
-      inhibitLearn();
-    }
+      return PROCESSED;
 
-    return PROCESSED;
+    case OPC_NERD:
+      // 57 - request for all stored events
+      DEBUG_SERIAL << F("> NERD : request all stored events for nn = ") << nn << endl;
 
-  case OPC_RQEVN:
-    // received RQEVN -- request for number of stored events
-    // DEBUG_SERIAL << F("> RQEVN -- number of stored events for nn = ") << nn << endl;
+      if (nn == module_config->nodeNum) 
+      {
+        msg->len = 8;
+        msg->data[0] = OPC_ENRSP;     // response opcode
+        msg->data[1] = highByte(nn);  // my NN hi
+        msg->data[2] = lowByte(nn);   // my NN lo
 
-    if (nn == module_config->nodeNum)
-    {
-      // respond with 0x74 NUMEV
+        for (byte i = 0; i < module_config->EE_MAX_EVENTS; i++) 
+        {
+          if (module_config->getEvTableEntry(i) != 0) 
+          {
+            // it's a valid stored event
+            // read the event data from EEPROM
+            // construct and send a ENRSP message
+            module_config->readEvent(i, &msg->data[3]);
+            msg->data[7] = i;  // event table index
 
-      controller->sendMessageWithNN(OPC_NUMEV, module_config->numEvents());
-    }
+            DEBUG_SERIAL << F("> sending ENRSP reply for event index = ") << i << endl;
+            controller->sendMessage(msg);
+            delay(10);
 
-    return PROCESSED;
+          }  // valid stored ev
+        }    // loop each ev
+      }      // for me
 
-  case OPC_NERD:
-    // request for all stored events
-    // DEBUG_SERIAL << F("> NERD : request all stored events for nn = ") << nn << endl;
+      return PROCESSED;
 
-    if (nn == module_config->nodeNum)
-    {
-      msg->len = 8;
-      msg->data[0] = OPC_ENRSP;                       // response opcode
-      msg->data[1] = highByte(module_config->nodeNum);        // my NN hi
-      msg->data[2] = lowByte(module_config->nodeNum);         // my NN lo
+    case OPC_NENRD:
```

Comment:
What about checking for short messages? You do that everywhere else. ;)

Reply:
The message is only opcode and node number, so if the message is short you don't know if it was for you or not.